graph_weather/models/atmorep/training/train.py

```diff
@@ -0,0 +1,392 @@
+
+    Returns:
+        A tuple (epoch_loss, epoch_field_losses) where:
+         - epoch_loss is the total loss summed over batches.
+         - epoch_field_losses is a dict with summed per-field losses.
+    """
+    model.train()
+    epoch_loss = 0.0
+    epoch_field_losses = {field: 0.0 for field in model.config.input_fields}
+
+    for batch_idx, batch_data in enumerate(dataloader):
+        # Generate training masks for the batch.
+        batch_masks = generate_training_masks(batch_data, model.config)
+
+        # Move data and masks to the device of the model.
+        device = next(model.parameters()).device
+        batch_data = {k: v.to(device) for k, v in batch_data.items()}
+        batch_masks = {k: v.to(device) for k, v in batch_masks.items()}
+
+        # Forward pass.
+        predictions = model(batch_data, batch_masks)
+
+        # Calculate loss.
+        loss, loss_components = loss_fn(predictions, batch_data, batch_masks)
+
+        # Backward pass and optimization.
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+        epoch_loss += loss.item()
+        for field, field_loss in loss_components["field_losses"].items():
+            epoch_field_losses[field] += field_loss
+
+    return epoch_loss, epoch_field_losses
+
+
+def _save_checkpoint(model, optimizer, config, output_dir, epoch, best_val_loss, is_best, history):
+    """
+    Save a training checkpoint.
+
+    Args:
+        model: The trained model.
+        optimizer: The optimizer.
+        config: The configuration object.
+        output_dir: Directory where the checkpoint should be saved.
+        epoch: Current epoch number.
+        best_val_loss: The best validation loss so far.
+        is_best: Boolean indicating if this is the best model.
+        history: Dictionary containing training history.
+    """
+    checkpoint = {
+        "epoch": epoch,
+        "model": model.state_dict(),
+        "optimizer": optimizer.state_dict(),
+        "best_val_loss": best_val_loss,
+        "is_best": is_best,
+        "history": history,
+        "config": config.__dict__ if hasattr(config, "__dict__") else config,
+    }
+    save_path = os.path.join(output_dir, f"checkpoint_epoch_{epoch}.pth")
+    torch.save(checkpoint, save_path)
+
+
+def train_atmorep(config, era5_path: str, output_dir: str, model=None, resume_from=None):
+    """
+    Train the AtmoRep model.
+
+    Args:
+        config: Configuration object with model parameters.
+        era5_path: Path to ERA5 data.
+        output_dir: Directory to save model checkpoints.
+        model: Optional pre-created model (if None, will be created).
+        resume_from: Optional path to checkpoint to resume training from.
+
+    Returns:
+        Trained model.
+    """
+    # Create model if not provided.
+    if model is None:
+        # Use absolute import so that the test patch is effective.
+        from graph_weather.models.atmorep.model.atmorep import AtmoRep
```

Comment:
This should be imported at the top of the file, don't include test specific logic in this library code. The tests should handle it.